### PR TITLE
cmake: Use a combined recipe that can download binaries or build from sources

### LINKED ISSUE #16607: [package] cmake/*: Separate recipes for from-source builds don't work in shared repos
### Description

#16503 changed the CMake recipe to download binaries from KitWare for macOS, Linux, and Windows, which is a welcome change. It also preserved the from-source recipes, which was also nice.

I've been working to get the new recipes up on our Solaris machine, and ran into a few problems. I can _make_ a CMake just fine, but...

- We can't generically say something like `cmake/[>=3.23.0]` in tool_requires, because that'll get the recipe for CMake 3.25.3 which is binary-only, and doesn't support Solaris. The source recipe version has 3.25.2.
- If there are two CMake recipes with the same version, but one downloads binaries, and one is source-only, they can't coexist in our Artifactory server, obviously, because the recipe IDs are different. Uploading packages built from source will obliterate packages built from binaries, and vice-versa.
- In our case, it specifically fails because the options are not the same either.

I think it would work a lot better if there were one recipe that figured out what to do based on the OS, or via an option or conf variable, and either downloaded binaries, or built from source.

Any other ideas?

/cc @jcar87 @datalogics-robb 

### Package and Environment Details

* Package Name/Version: **CMake/3.25**
* Operating System+version: **Solaris**
* Compiler+version: **GCC 9**
* Docker image: **conanio/gcc8**
* Conan version: **conan 1.59.0**
* Python version: **Python 3.9.16**

### Conan profile

[settings]
arch=sparcv9
arch_build=sparcv9
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=9.2
os=SunOS
os_build=SunOS
doxygen:arch=sparcv9
doxygen:build_type=Release
cmake:arch=sparcv9
cmake:build_type=Release
ninja:arch=sparcv9
ninja:build_type=Release
swig:arch=sparcv9
swig:build_type=Release
b2:arch=sparcv9
b2:build_type=Release
innoextract:arch=sparcv9
innoextract:build_type=Release
m4:build_type=Release
m4:arch=sparcv9
autoconf:build_type=Release
autoconf:arch=sparcv9
automake:build_type=Release
automake:arch=sparcv9
[options]
b2:toolset=cxx
b2:use_cxx_env=True
boost:debug_level=1
cmake:with_openssl=False
pcre:with_bzip2=False
[build_requires]
*: cmake/[>=3.23.0]
[env]
CC=/opt/gcc-9.2.0/bin/gcc
CFLAGS=-pthread -m64
CXX=/opt/gcc-9.2.0/bin/g++
CXXFLAGS=-pthread -m64
LDFLAGS=-Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
PATH=[/opt/cmake-3.19/bin,/opt/gcc-9.2.0/bin,/usr/gnu/bin]
b2:CXXFLAGS=-pthread -m64 -Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
b2:LD_RUN_PATH=/opt/gcc-9.2.0/lib
m4:CFLAGS=-pthread -m64 -std=gnu99
[buildenv]
CC=/opt/gcc-9.2.0/bin/gcc
CXX=/opt/gcc-9.2.0/bin/g++
CXXFLAGS=-pthread -m64
LDFLAGS=-Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
PATH=+(path)/opt/gcc-9.2.0/bin:/usr/gnu/bin
PATH+=(path)/opt/cmake-3.19/bin
CFLAGS=-pthread -m64
m4:CFLAGS="-pthread -m64 -std=gnu99"
b2:CXXFLAGS="-pthread -m64 -Wl,-R,/opt/gcc-9.2.0/lib/sparcv9"
b2:LD_RUN_PATH=/opt/gcc-9.2.0/lib

### Steps to reproduce

conan create recipes/swig/dl swig/4.0.2+dl.2@ --update --json /tmp/pytest-of-devauto/pytest-7/test_build_tool_swig_4_0_2_dl_0/create.json --profile:host build-profile-solaris-sparc --profile:host profiles/cmake-tool --options:host pcre:with_bzip2=False --build missing

### Logs

<details><summary>Click to expand log</summary>

```
Creating package swig/4.0.2+dl.2: conan create recipes/swig/dl swig/4.0.2+dl.2@ --update --json /tmp/pytest-of-devauto/pytest-7/test_build_tool_swig_4_0_2_dl_0/create.json --profile:host build-profile-solaris-sparc --profile:host profiles/cmake-tool --options:host pcre:with_bzip2=False --build missing
************************************************************
  This package defines both 'os' and 'os_build' 
  Please use 'os' for libraries and 'os_build'
  only for build-requires used for cross-building
************************************************************
Exporting package recipe
swig/4.0.2+dl.2 exports: File 'conandata.yml' found. Exporting it...
swig/4.0.2+dl.2 exports: Copied 1 '.yml' file: conandata.yml
swig/4.0.2+dl.2 exports_sources: Copied 4 '.patch' files: 0002-4.0.2-do-not-define-SWIG_LIB_WIN_UNIX.patch, 0004-1.3.40-swig-linux-library-path.patch, 0003-1.3.40-do-not-define-SWIG_LIB_WIN_UNIX.patch, 0001-swig-linux-library-path.patch
swig/4.0.2+dl.2 exports_sources: Copied 1 '.cmake' file: conan-official-swig-targets.cmake
swig/4.0.2+dl.2: The stored package has not changed
swig/4.0.2+dl.2: Using the exported files summary hash as the recipe revision: 1bdf819c2d33a17c1f4f222ac87a3b27 
swig/4.0.2+dl.2: Exported revision: 1bdf819c2d33a17c1f4f222ac87a3b27
Configuration:
[settings]
arch=sparcv9
arch_build=sparcv9
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=9.2
os=SunOS
os_build=SunOS
doxygen:arch=sparcv9
doxygen:build_type=Release
cmake:arch=sparcv9
cmake:build_type=Release
ninja:arch=sparcv9
ninja:build_type=Release
swig:arch=sparcv9
swig:build_type=Release
b2:arch=sparcv9
b2:build_type=Release
innoextract:arch=sparcv9
innoextract:build_type=Release
m4:build_type=Release
m4:arch=sparcv9
autoconf:build_type=Release
autoconf:arch=sparcv9
automake:build_type=Release
automake:arch=sparcv9
[options]
b2:toolset=cxx
b2:use_cxx_env=True
boost:debug_level=1
cmake:with_openssl=False
pcre:with_bzip2=False
[build_requires]
*: cmake/[>=3.23.0]
[env]
CC=/opt/gcc-9.2.0/bin/gcc
CFLAGS=-pthread -m64
CXX=/opt/gcc-9.2.0/bin/g++
CXXFLAGS=-pthread -m64
LDFLAGS=-Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
PATH=[/opt/cmake-3.19/bin,/opt/gcc-9.2.0/bin,/usr/gnu/bin]
b2:CXXFLAGS=-pthread -m64 -Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
b2:LD_RUN_PATH=/opt/gcc-9.2.0/lib
m4:CFLAGS=-pthread -m64 -std=gnu99
[buildenv]
CC=/opt/gcc-9.2.0/bin/gcc
CXX=/opt/gcc-9.2.0/bin/g++
CXXFLAGS=-pthread -m64
LDFLAGS=-Wl,-R,/opt/gcc-9.2.0/lib/sparcv9
PATH=+(path)/opt/gcc-9.2.0/bin:/usr/gnu/bin
PATH+=(path)/opt/cmake-3.19/bin
CFLAGS=-pthread -m64
m4:CFLAGS="-pthread -m64 -std=gnu99"
b2:CXXFLAGS="-pthread -m64 -Wl,-R,/opt/gcc-9.2.0/lib/sparcv9"
b2:LD_RUN_PATH=/opt/gcc-9.2.0/lib


JSON file created at '/tmp/pytest-of-devauto/pytest-7/test_build_tool_swig_4_0_2_dl_0/create.json'
ERROR: cmake/3.25.3: option 'with_openssl' doesn't exist
Possible options are []
```

</details>

### PR DESCRIPTION
**Pull request by datalogics-kam**
Specify library name and version:  **cmake/***

Fixes #16607

Having a separate recipe for CMake from sources isn't compatible with placing CMake recipes into the same remote repo. The recipe revisions keep overwriting each other. Therefore, I combined the recipes.

This solves our problem for Solaris. We have it in our fork, and I'm offering it to the community.

- By default, uses binaries on macOS, Windows and Linux.
- `from_sources` option can be used to build from sources on other platforms, or to force sources on Windows or Linux.
- The ability to bootstrap CMake was removed. If you need a CMake to build CMake from sources, either use a CMake that is already in Conan, or build a CMake with autotools by hand.

I do get
```
[HOOK - conan-center.py] pre_export(): ERROR: [CONANDATA.YML FORMAT (KB-H030)] First level entries ['binaries'] not allowed. Use only first level entries ['sources', 'patches'] in conandata.yml (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H030-CONANDATA.YML-FORMAT)
```

Suggestions welcome on how best to segregate data for sources and binaries in `conandata.yml` to conform to the requirements of the hooks.

I can also reorganize the recipe folder as requested.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
> Suggestions welcome on how best to segregate data for sources and binaries in `conandata.yml` to conform to the requirements of the hooks.

It's not just the hooks it's also the built-in client helpers :)

You should do it like

```yml
"sources":
    "3.25.3":
        "code":
             // ... url, sha
        "precompiled":
             //...
```

**Comment by jcar87** [comment]
Hi @datalogics-kam  - thanks for your contribution. We'll take this into consideration as I can see how recipe revisions overwriting each other can cause the problems you describe.

I have a couple of questions:

> The ability to bootstrap CMake was removed. If you need a CMake to build CMake from sources, either use a CMake that is already in Conan, or build a CMake with autotools by hand.

If having a functional CMake is already required to build cmake from source, and this new combined recipe does _not_ depend on cmake as a `tool_requires` - then this would imply that there is already a functional CMake available in the system, I'm assuming Solaris as well?

Generally speak, we prefer recipes in Conan Center to _not_ depend on CMake, and assume that up to a certain version it is available on the system - this has proven problematic over time as libraries have a requirement for newer versions. However, this is now causing a situation where a single `conan install` with `--build=missing` may actually require multiple versions of CMake to be available and built, when in practice a single version (the most recent) would suffice.

For this, we have been considering the combination of two different approaches: firstly, the ability to _conditionally_ depend on CMake as a `tool_requires` if a suitable version is not available on the system (in a more formal way than it is done in some of the recipe), and  secondly, the use of version ranges. Because even if CMake was buildable from source, we want to avoid a situation where any user has to spend time _compiling_ multiple version of CMake for any reason.

Would addressing these in the manner described above be a suitable solution to your problem? that is, have it in such a way that if the system version of CMake in your Solaris setup is already a suitable version (outside of Conan) - to skip the dependency altogether? 
Obviously we may still want a recipe that can build it from source, when the version isn't suitable.

**Comment by datalogics-kam** [comment]
> If having a functional CMake is already required to build cmake from source, and this new combined recipe does not depend on cmake as a tool_requires - then this would imply that there is already a functional CMake available in the system, I'm assuming Solaris as well?

I took the tack that when in the rare instance of building CMake from sources, the end user has to provide a CMake that will do that. At DL, _our_ automated tool building job takes care of that by providing CMake in a `[tool_requires]` in a profile. We keep that CMake-for-building-CMake in Artifactory as `bootstrap_cmake` (we made that one using CMAKE_RECIPE_NAME) and are careful not to break it. It works well, because we can build any current version of CMake with any CMake>=3.15.

I took out the bootstrap support because anybody who really needs it already has to know what they're doing. If they have to build a CMake to get going from scratch, they can follow the instructions from Kitware, make a CMake, and use that with this recipe to make an initial Conan package. We weren't even in that situation, so I figured it was less code to maintain. #16503 said the bootstrap wasn't even working, anyway.

> Generally speak, we prefer recipes in Conan Center to not depend on CMake, and assume that up to a certain version it is available on the system - this has proven problematic over time as libraries have a requirement for newer versions. However, this is now causing a situation where a single conan install with --build=missing may actually require multiple versions of CMake to be available and built, when in practice a single version (the most recent) would suffice.

I'm not proposing that any recipe in CCI have a `conanfile.py` requirement for CMake, if that's what you're concerned about. We make CMake available in our company by putting it in `[tool_requires]` in a profile.

> For this, we have been considering the combination of two different approaches: firstly, the ability to conditionally depend on CMake as a tool_requires if a suitable version is not available on the system (in a more formal way than it is done in some of the recipe), and secondly, the use of version ranges. Because even if CMake was buildable from source, we want to avoid a situation where any user has to spend time compiling multiple version of CMake for any reason.

The "conditionally depend" part, is that a proposed new Conan feature that would check the local OS before downloading/building a tool? Sounds useful, even if my company not be the target audience for it.

We use version ranges for our `[tool_requires]` (in profiles) for CMake now, as `cmake@[>=3.23.0]`. We have our own remote fed by a fork of this repository, so we have a lot of control over what's in it, so we just use "the latest CMake that's passed our tests and is in Artifactory" as it were.

> Would addressing these in the manner described above be a suitable solution to your problem? that is, have it in such a way that if the system version of CMake in your Solaris setup is already a suitable version (outside of Conan) - to skip the dependency altogether?
Obviously we may still want a recipe that can build it from source, when the version isn't suitable.

We don't plan on having a system version of CMake in Solaris any more.

Our specially-named `bootstrap_cmake` is available to build the actual `cmake`, allowing us to use `tool_requires` to specify a CMake version. We're getting away from having CMake installed on our build systems, as part of using every tool we can from Conan. Our goal is that only two things need to be preinstalled: Python, and the build tools/compilers.

Thanks for getting back to me. I hope I addressed your questions and concerns; if not, please comment further and I will do my best.

**Comment by datalogics-kam** [comment]
```
======== Installing packages ========
cmake/3.22.6: Calling source() in /home/conan/w/prod-v2/BuildSingleReference/p/cmakeffaa55e1f314b/s
ERROR: cmake/3.22.6: Error in source() method, line 158
	if not self.options.from_sources:
	ConanException: 'self.options' access in 'source()' method is forbidden
```
This is a roadblock...how am I supposed to have it able to configure to build from sources or not, if options are forbidden in the `sources()` method?

Could I use a `conf` variable or something? Are those allowed in `sources()`? Are there per-package `conf` settings?

Looking at the source code for Conan, it looks like options and settings are prohibited in `sources()`, so that rules out going by `settings.os` as well.

I'm going to appeal to the Conan experts here...what's the best way to handle this situation, so we can have a recipe that can both use binaries on platforms that have them, and use sources otherwise? Should this be a special condition we can turn on in Conan v2, to allow more flexibility?

**Comment by datalogics-kam** [comment]
Also, advice on the hook warnings in [this comment](https://github.com/conan-io/conan-center-index/pull/17032#issuecomment-1517019208)?

* I can't really help the path lengths in `CMake.app`.
* Feels strange setting `cpp_info.frameworks` for a tool...is this something I'm missing in the recipe config?

**Comment by datalogics-kam** [comment]
Noting that the hook warnings in [this comment](https://github.com/conan-io/conan-center-index/pull/17032#issuecomment-1517019208) are similar to the ones in #17307, so I'm not going to worry about them.

**Comment by datalogics-kam** [comment]
I'm going to close and open this to see if I can get the CI to run. No results after 5 days.

**Comment by prince-chrismc** [comment]
> This is a roadblock...how am I supposed to have it able to configure to build from sources or not, if options are forbidden in the sources() method?

You need to download binaries in the build method :)

**Comment by datalogics-kam** [comment]
> You need to download binaries in the build method :)

I ended up getting the sources in the `generate()` method. The point of this change is to have a recipe that can use binaries on platforms that have them, and sources on the others.

I tested this thoroughly locally, with Conan 1 and 2, and with all the linters, and it's all happy...the question is what can I do to get the CI to finish running here?

**Comment by prince-chrismc** [comment]
Fix the merge conflicts? :)

**Comment by datalogics-kam** [comment]
> Fix the merge conflicts? :)

Oh, right, didn't notice it was being tested in merge mode.  Thanks.

**Comment by prince-chrismc** [comment]
Ignore the yaml linter, its not required

**Comment by datalogics-kam** [comment]
Can someone explain about how I can diagnose and fix the Conan 2 build, and which of the warnings have to be fixed? Thanks!

File: recipes/cmake/combined/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        "from_sources": [True, False],
+    }
+    default_options = {
+        "with_openssl": True,
+        "from_sources": False,
+    }
+
+    def set_name(self):
+        self.name = os.environ.get('CMAKE_RECIPE_NAME', 'cmake')
+        print(self.name)
```
Comment (Hopobcn) [inline recipes/cmake/combined/conanfile.py:35]:
```suggestion
```
Is this required?

Comment (prince-chrismc) [reply recipes/cmake/combined/conanfile.py:35]: This is a bit out of scope for the role of ConanCenter. This is a very niche problem which the majority of users do not face so it's hard to justify the additional complexity.

CMake being such a well known tool, it's often used as an example and this is not something we'd recommend. 

> if we somehow build a broken one and store it

I would recommend improving your testing, the `--test-folder` is there to give you this flexibility, as a first step rather then having a separate name just to

> we don't have to do it again

This is was lockfiles are perfect for so I'd encourage you to look at your process :)

Comment (datalogics-kam) [reply recipes/cmake/combined/conanfile.py:35]: I'll take `set_name` out then, no biggie. 

> I would recommend improving your testing

True, we don't _upload_ anything that's broken...except that if we say "use the latest cmake as a `tool_requires` to build cmake", then build a broken one, it's in the _local_ cache, fails tests, tell the automation to build it again, and it tries to build CMake with a version of itself that it just built that doesn't work...It's a bootstrapping problem, and you need something stable if the "build yourself with yourself" loop fails.

CMake is unusual in the sense that it's so critical to building almost everything else, and is best built with another copy of itself.

But your commentary makes me think maybe I can rethink the process...it's amazing what fresh insights come several months after the release of solution 1.x. Thanks.

(As an aside, I'm _so old_, that I remember getting GCC from a series of postings on `alt.sources.something` on Usenet, compiling it with the K&R compiler on a machine, then compiling it with itself, then compiling it with itself for a second time, checking to make sure the second and third builds produce the same result. The analogue of the K&R compiler is building the Conan CMake package with a CMake that was bootstrapped with autotools, or a carefully stored "known good" Conan CMake package.)

**Review by Hopobcn — APPROVED** [review]

File: recipes/cmake/combined/conanfile.py (hunk 1 above)
Comment (prince-chrismc) [reply recipes/cmake/combined/conanfile.py]: ❤️

**Comment by datalogics-kam** [comment]
I'm not going 

> Hooks produced the following warnings for commit [817aa3f](https://github.com/conan-io/conan-center-index/commit/817aa3fa3bbf838a4cf731653a47c71421f922bf)

I'm not going to worry about these, because [similar warnings appeared](https://github.com/conan-io/conan-center-index/pull/19064#issuecomment-1666804666) in #19064, and didn't seem to impede the approval.

**Comment by datalogics-kam** [comment]
I'd press **Update branch**, but there are no newer changes to the CMake recipe in `master`, and I don't what to thrash the build queue.

Once there are enough approvals, perhaps?

**Pull request closed without merging**